File: quickstart_guides/sorting_searching/python/find_fixed_point.py
"""
Title: Find fixed point

Problem:
    A fixed point in an array "A" is an index "i" such that A[i] is equal to
    "i".  Given an array of n distinct integers sorted in ascending order,
    write a function that returns a "fixed point" in the array. If there is not
    a fixed point return "None".

Execution: python find_fixed_point.py
"""
from typing import List, Union
import unittest


# Time Complexity: O(log n)
# Space Complexity: O(1)
def find_fixed_point(input_list: List[int]) -> Union[int, None]:
    low = 0
    high = len(input_list) - 1

    while low <= high:
        mid = (low + high) // 2

        if input_list[mid] < mid:
            low = mid + 1
        elif input_list[mid] > mid:
            high = mid - 1
        else:
            return input_list[mid]
    return None


class TestFindFixedPoint(unittest.TestCase):
    """Unit tests for find_fixed_point."""

    def test_1(self):
        input_list = [-10, -5, 0, 3, 7]
        self.assertEqual(find_fixed_point(input_list), 3)

    def test_2(self):
        input_list = [0, 2, 5, 8, 17]
        self.assertEqual(find_fixed_point(input_list), 0)

    def test_3(self):
        input_list = [-10, -5, 3, 4, 7, 9]
        self.assertEqual(find_fixed_point(input_list), None)


if __name__ == "__main__":
    unittest.main()

File: quickstart_guides/sorting_searching/java/FindFixedPoint.java
/*
 *   Title: Find fixed point
 *
 *   Problem:
 *      Given an array of n duplicates or distinct integers sorted in ascending
 *      order, write a function that returns a Fixed Point in the array, if
 *      there is any Fixed Point present in the array, else returns -1. Fixed
 *      Point in an array is an index i such that arr[i] is equal to i. Note
 *      that integers in the array can be negative.
 *
 *   Execution: javac FindFixedPoint.java && java FindFixedPoint
 */
class FindFixedPoint
{
	// Main Function to find fixed
	// index using binary search
	public static int findFixedPoint(int arr[], int low,
									int high)
	{
		if (high < low)
			return -1;

		// low + (high - low) / 2
		int mid = (low + high) / 2;
		int midValue = arr[mid];

		if (mid == arr[mid])
			return mid;

		// Search left.
		int leftindex = Math.min(mid - 1, midValue);
		int left = findFixedPoint(arr, low, leftindex);

		if (left >= 0)
			return left;

		// Search right.
		int rightindex = Math.max(mid + 1, midValue);
		int right = findFixedPoint(arr, rightindex, high);

		return right;
	}

	public static void main(String[] args)
	{
		// input 1
		int arr[] = {-10, -5, 2, 2, 2,
					3, 4, 7, 9, 12, 13};

        assert findFixedPoint(arr, 0, arr.length - 1) == 2;

		// input 2
		int arr1[] = {-10, -1, 3, 3, 10,
						30, 30, 50, 100};

        assert findFixedPoint(arr1, 0, arr1.length - 1) == 3;

		System.out.println("All tests passed.");
	}
}
